processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -512,26 +512,27 @@ private long getMaxTimeMillis()
     return getFacts().lastKey().getTimestamp();
   }
 
-  private String[] getDimVals(final DimDim dimLookup, final List<String> dimValues)
+  private int[] getDimVals(final DimDim dimLookup, final List<String> dimValues)
   {
-    final String[] retVal = new String[dimValues.size()];
     if (dimValues.size() == 0) {
       // NULL VALUE
-      if (!dimLookup.contains(null)) {
-        dimLookup.add(null);
-      }
+      dimLookup.add(null);
       return null;
     }
+
+    if (dimValues.size() == 1) {
+      return new int[]{dimLookup.add(dimValues.get(0))};
+    }
+
+    String[] dimArray = dimValues.toArray(new String[dimValues.size()]);
+    Arrays.sort(dimArray);
```

Comment:
hmmm.... but, `get(int index)` is part of List interface contract. It will be very weird for someone to implement an InputRow which returns a dimension values List that does not implement get(int) . (in that case actually that InputRow implementation should be fixed).